### PR TITLE
fix(icos): Several testing and performance improvements to SetupOS

### PR DESCRIPTION
**Pull request by DFINITYManu**
This commit encompasses two changes:

## Improvements in IC OS testing

* Improving performance of HostOS and GuestOS installation within tests and in prod.
* Document how to start SetupOS interactively.
* Provide facilities for interactive testing unobstructed by automated setup process.
* Change methodology of SetupOS testing to speed up tests.

## Interactive testing of SetupOS

This new testing feature permits the execution of a SetupOS installer in two modes:

1. Automatic install then reboot to HostOS.
2. Interactive debug without install.

This only works in the dev environment established by Bazel.

This feat is accomplished by the addition of the following variants:

* //ic-os/setupos/envs/dev:interactive-install for installation (with
-no-kvm for scenarios where no acceleration exists)
* //ic-os/setupos/envs/dev:interactive-debug for no-install inspection
of SetupOS (same -no-kvm availability)

To enable this to work with higher performance, we create a new disk
image variant that does not tarball the SetupOS image.

Copying the image to the temporary directory is near instantaneous since
various kernel optimizations at play using `cp`
kick in that do not kick in when using `tar x`.

Total time in my ThinkPad (without KVM acceleration) for the whole
install + reboot + boot to HostOS + login + poweroff is:
```
real    7m31.450s
user    6m16.889s
sys     0m17.826s
```

The change deletes the `launch-*` targets in the dev environment of
SetupOS, since these targets are non-sensical to begin with (the
system boots but promptly fails). The targets outlined above provide a
superior substitute to the targets being deleted here.

### CODE REVIEW

File: ic-os/components/setupos-scripts/check-network.sh
Comment (andrewbattat) [inline ic-os/components/setupos-scripts/check-network.sh:222]: To me, it would make more sense to have all the parameters be *skip* booleans so then the logic would go like:

if kernel_cmdline_bool ic.setupos.skip_network_checks; then
     echo "* Network checks skipped by request via kernel command line"
else
     read_variables
     ...

It's confusing to me that currently in the mainnet path, ic.setupos.check_network is absent, but kernel_cmdline_bool_default_true defaults to true, and therefore we do check network

If they're all "skip" parameters instead, we don't even need the "default" logic. We just return true if the parameter is set and else return false

File: ic-os/components/setupos-scripts/install-guestos.sh
Comment (andrewbattat) [inline ic-os/components/setupos-scripts/install-guestos.sh:30]: Is this comment all necessary? It's good information but do we need it in code? Especially when it's referring to "before"

File: ic-os/setupos/README.adoc
Comment (andrewbattat) [inline ic-os/setupos/README.adoc:27]:
```suggestion
Bazel can be used to boot a local VM with a SetupOS like this:
```

**Comment by andrewbattat** [comment]
Thank you for doing this Manuel!

File: ic-os/components/setupos-scripts/check-hardware.sh
Comment (andrewbattat) [inline ic-os/components/setupos-scripts/check-hardware.sh:1]: One thing I'm paranoid about is a bug being introduced where this if statement resolves to false on production, and then we are skipping checks by default. With the old way of doing things, this was not a risk as we didn't have a conditional directly in SetupOS.

File: ic-os/components/setupos-scripts/install-guestos.sh
Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/install-guestos.sh:30]: Given that the repo squashes individual commits, so I can't put this in the commit message, it's useful information for future people who need to find out why this got changed to be the way it is.

File: ic-os/components/setupos-scripts/functions.sh
Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/functions.sh:108]: Very cool, but that doesn't parse quoted strings in the command line (which is valid input).

I've modified it slightly to do this, and will incorporate that to the parent PR.

File: ic-os/components/setupos-scripts/check-network.sh
Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/check-network.sh:222]: I don't like that idea because it requires people reading the code to think in terms of double negatives.  Having worked in the Dracut initial RAM disk generator myself, they moved away from instituting new parameters that would behave this way because it made the code very difficult to understand.  Of course, the existing parameters stayed for backwards compat reasons.

Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/check-network.sh:222]: > It's confusing to me that currently in the mainnet path, ic.setupos.check_network is absent, but kernel_cmdline_bool_default_true defaults to true, and therefore we do check network

Not sure I understand this.

```
    if kernel_cmdline_bool_default_true ic.setupos.check_network; then
        read_variables
        get_network_settings
        print_network_settings

        if [[ -n ${ipv4_address} && -n ${ipv4_prefix_length} && -n ${ipv4_gateway} ]]; then
            validate_domain_name
            setup_ipv4_network
            ping_ipv4_gateway
        fi

        ping_ipv6_gateway
        assemble_nns_nodes_list
        query_nns_nodes
    else
        echo "* Network checks skipped by request via kernel command line"
    fi
```

Is it not clear that we're asking if the parameter is true or is absent, we must check?  In other words, to check **is** the default.

File: ic-os/components/setupos-scripts/functions.sh
Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/functions.sh]: The incorporation has been incorporated.  Thanks!

File: ic-os/components/setupos-scripts/check-network.sh
Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/check-network.sh:222]: Maybe we can name the conditional function better?  Something like `boot_flag_or_true`?

File: ic-os/defs.bzl
Comment (Bownairo) [inline ic-os/defs.bzl:568]: https://github.com/dfinity/ic/pull/1451#discussion_r1786898444

> This same tooling is used for testing HostOS and GuestOS as well. If you'd like to add this for testing SetupOS, could you add a branch to the logic? We have a similar thing above for the remote VMs, in setting the `nested` flag.

File: ic-os/components/setupos-scripts/functions.sh
Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/functions.sh:82]: Yes.

File: ic-os/components/setupos-scripts/functions.sh
Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/functions.sh:107]: Yes.

File: ic-os/components/setupos-scripts/install-guestos.sh
Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/install-guestos.sh:21]: Nothing.  It's going to be deleted in the next push.  You were right!

File: ic-os/components/setupos-scripts/install-guestos.sh
Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/install-guestos.sh:30]: Yes.

File: ic-os/defs.bzl
Comment (DFINITYManu) [reply ic-os/defs.bzl:566]: You are right.

File: ic-os/components/setupos-scripts/check-network.sh
Comment (andrewbattat) [reply ic-os/components/setupos-scripts/check-network.sh:222]: @Bownairo WDYT? I'll defer to you

File: ic-os/components/setupos-scripts/check-hardware.sh
Comment (andrewbattat) [reply ic-os/components/setupos-scripts/check-hardware.sh:1]: @Bownairo WDYT? I'll defer to you

File: ic-os/components/setupos-scripts/check-network.sh
Comment (Bownairo) [reply ic-os/components/setupos-scripts/check-network.sh:222]: It doesn't matter to me much which direction this goes, but to @andrewbattat's point I do think it could be simpler to stick with one "default".

The variables we have now are:
```
check_hardware (default true)
check_network (default true)
check_age (default true)
stop_before_installation (default false)
stop_before_reboot (default false)
```

If we pick `false` as default:
```
skip_hardware_check (default false)
skip_network_check (default false)
skip_age_check (default false)
stop_before_installation (default false)
stop_before_reboot (default false)
```

If we pick `true` as default:
```
check_hardware (default true)
check_network (default true)
check_age (default true)
perform_installation (default true)
reboot_after_installation (default true)
```

File: ic-os/components/setupos-scripts/check-hardware.sh
Comment (Bownairo) [reply ic-os/components/setupos-scripts/check-hardware.sh:1]: 🔥🥔

@slawomirbabicz this is OK with me if it is OK with you.

File: ic-os/defs.bzl
Comment (Bownairo) [reply ic-os/defs.bzl:566]: Could you make this change?

File: ic-os/defs.bzl
Comment (Bownairo) [reply ic-os/defs.bzl:568]: Node uses this to test GuestOS and HostOS, and this change breaks:

```
% bazel run ic-os/guestos/envs/dev:launch-local-vm-no-kvm

...

Gave up waiting for root file system device.  Common problems:
 - Boot args (cat /proc/cmdline)
   - Check rootdelay= (did the system wait long enough?)
 - Missing modules (cat /proc/modules; ls /dev)
ALERT!  /dev/vda5 does not exist.  Dropping to a shell!


BusyBox v1.36.1 (Ubuntu 1:1.36.1-6ubuntu3.1) built-in shell (ash)
Enter 'help' for a list of built-in commands.

(initramfs)
```

File: ic-os/components/BUILD.bazel
Comment (Bownairo) [inline ic-os/components/BUILD.bazel:54]: These scripts should not be excluded, was the new `BUILD.bazel` causing issues? It should have been captured by `**/BUILD.bazel`.

**Review by Bownairo — COMMENTED** [review]
> Alright! I think we're ready to go. What is missing from this PR?

Just a few last comments! If @slawomirbabicz could review, I will move to approve as well.

File: ic-os/components/setupos-scripts/install-hostos.sh
Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/install-hostos.sh:26]: Happily.  Please stand by.

File: ic-os/defs.bzl
Comment (DFINITYManu) [reply ic-os/defs.bzl]: Yes.  Done.

File: ic-os/defs.bzl
Comment (DFINITYManu) [reply ic-os/defs.bzl:568]: I can see why.  I will fix very shortly.

File: ic-os/components/BUILD.bazel
Comment (Bownairo) [reply ic-os/components/BUILD.bazel:54]: Sure! What was the problem? I see the tests all passing with this change removed.

```diff
% git diff
diff --git a/ic-os/components/BUILD.bazel b/ic-os/components/BUILD.bazel
index 2e8d083e82..1e05b0f825 100644
--- a/ic-os/components/BUILD.bazel
+++ b/ic-os/components/BUILD.bazel
@@ -51,7 +51,6 @@ REPO_COMPONENTS = glob(
         "**/*.bzl",
         "**/*.md",
         "**/*.adoc",
-        "setupos-scripts/*",
     ],
 )
```

```
% bazel clean --expunge
INFO: Invocation ID: e6782ac9-c158-4203-909d-341953acd714
INFO: Starting clean (this may take a while). Consider using --async if the clean takes more than several minutes.

% bazel test ic-os/components/...
Starting local Bazel server and connecting to it...
INFO: Invocation ID: d4748570-d159-4ef7-848c-2396093eebd6

...

INFO: Analyzed 3 targets (206 packages loaded, 1727 targets configured).
INFO: Found 1 target and 2 test targets...
INFO: Elapsed time: 36.935s, Critical Path: 0.30s
INFO: 13 processes: 7 internal, 4 linux-sandbox, 2 local.
INFO: Build completed successfully, 13 total actions
//ic-os/components:check_unused_components_test                          PASSED in 0.0s
//ic-os/components/setupos-scripts:test-kernel-cmdline-function          PASSED in 0.1s

Executed 2 out of 2 tests: 2 tests pass.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
```

File: ic-os/defs.bzl
Comment (Bownairo) [inline ic-os/defs.bzl:560]: nit: This `cd` is no longer necessary with the full path.

File: ic-os/defs.bzl
Comment (DFINITYManu) [reply ic-os/defs.bzl:568]: I believe I have successfully fixed the launch-local-vm targets for GuestOS and HostOS.  See commit 2e80d280d29931f4e57e08de20907ee9487d2daf .

File: ic-os/components/BUILD.bazel
Comment (DFINITYManu) [reply ic-os/components/BUILD.bazel]: OK.  removed the exclude.

File: ic-os/components/setupos-scripts/check-network.sh
Comment (DFINITYManu) [reply ic-os/components/setupos-scripts/check-network.sh:222]: I picked True as default.  Good catch!  Thanks.  I've pushed changes already.

File: ic-os/defs.bzl
Comment (DFINITYManu) [reply ic-os/defs.bzl:560]: Thanks, but the next PR on top of this one already fixes this issue.

File: ic-os/components/setupos-scripts/check-network.sh
Comment (andrewbattat) [reply ic-os/components/setupos-scripts/check-network.sh:222]: Can we now delete kernel_cmdline_bool_default_false if it's no longer in use?

**Review by Bownairo — APPROVED** [review]

File: ic-os/defs.bzl
Comment (Bownairo) [reply ic-os/defs.bzl:560]: In general I think we should not merge any code to master that is missing fixes, but this case is just a nit.

**Comment by DFINITYManu** [comment]
Alright!  We're ready for review.  The `interactivetesting` branch was merged here, as requested in the other branch.

**Comment by gregorydemay** [comment]
Removing cross-chain team since the label was due to the PR being out-of-date with master.

**Comment by frankdavid** [comment]
> Alright! We're ready for review. The `interactivetesting` branch was merged here, as requested in the other branch.

Could you please still address the remaining comments in that PR so that I don't have to add them here again?

**Comment by DFINITYManu** [comment]
> > Alright! We're ready for review. The `interactivetesting` branch was merged here, as requested in the other branch.
> 
> Could you please still address the remaining comments in that PR so that I don't have to add them here again?

Yes.

I see this one:

https://github.com/dfinity/ic/pull/2115/#discussion_r1829241222

I'm not convinced about these suggested efficiency improvements of using String over Vec<char> since the smallest string is going to be more than 4 bytes anyway (remember it's not just the actual string stored but also the length) and we're going character by character.

I'm working on figuring out exactly how to implement what you suggested without getting errors, nonetheless.

**Comment by frankdavid** [comment]
> Not sure what that means.

That was just a recommendation to call the argument `disable_setup_checks` (as in `disable` is more commonly used than `defeat`).

**Review by frankdavid — APPROVED** [review]

File: ic-os/defs.bzl
Comment (Bownairo) [inline ic-os/defs.bzl:346]: Are these OK to put into the remote cache?

```
2.5G    ic-os/setupos/envs/dev/disk.img
1.6G    ic-os/hostos/envs/dev/disk.img
1.3G    ic-os/guestos/envs/dev/disk.img
```

**Review by Bownairo — APPROVED** [review]

File: .github/CODEOWNERS
Comment (DFINITYManu) [reply .github/CODEOWNERS]: Done.

File: ic-os/defs.bzl
Comment (DFINITYManu) [reply ic-os/defs.bzl:346]: We don't need to do that -- how do I disable it?

File: ic-os/defs.bzl
Comment (DFINITYManu) [reply ic-os/defs.bzl:606]: If I remove this, then this breaks the targets defined below.

Furthermore, who says this _can't_ this be used directly?  I use it directly all the time with arguments to test this code.

File: ic-os/defs.bzl
Comment (DFINITYManu) [reply ic-os/defs.bzl]: I turned it off.

**Pull request merged**
